### PR TITLE
ipn/ipnlocal: logout ephemeral nodes on shutdown

### PR DESCRIPTION
**Pull request by maisem**
Signed-off-by: Maisem Ali <maisem@tailscale.com>

### CODE REVIEW

File: ipn/ipnlocal/local.go [hunk 1]
```diff
@@ -342,7 +342,17 @@ func (b *LocalBackend) onHealthChange(sys health.Subsystem, err error) {
...
 	b.mu.Lock()
+	if b.shutdownCalled {
+		b.mu.Unlock()
+		return
+	}
 	b.shutdownCalled = true
+
+	if b.loginFlags&controlclient.LoginEphemeral != 0 {
+		b.mu.Unlock()
+		b.logout(b.ctx, true) // best effort
```
Comment (bradfitz) [inline ipn/ipnlocal/local.go:353]: can we bound the time here? Or start it async in its own goroutine?

Not a fan of blocking Shutdown with network work.

File: ipn/ipnlocal/local.go [hunk 2]
```diff
@@ -342,7 +342,17 @@ func (b *LocalBackend) onHealthChange(sys health.Subsystem, err error) {
...
+		b.mu.Lock()
```
Comment (bradfitz) [inline ipn/ipnlocal/local.go:354]: any invariants you need to re-check now that you've dropped the lock and reacquired it?

File: ipn/ipnlocal/local.go (hunk 1 above)
Comment (maisem) [reply ipn/ipnlocal/local.go:353]: Time bound sgtm
Running a different goroutine doesn't help shutdown

File: ipn/ipnlocal/local.go (hunk 2 above)
Comment (maisem) [reply ipn/ipnlocal/local.go:354]: Nothing can restart the backend, so no.

**Review by bradfitz — APPROVED** [review]

File: ipn/ipnlocal/local.go [hunk 3]
```diff
@@ -342,7 +342,19 @@ func (b *LocalBackend) onHealthChange(sys health.Subsystem, err error) {
...
+		b.mu.Unlock()
+		return
+	}
 	b.shutdownCalled = true
+
+	if b.loginFlags&controlclient.LoginEphemeral != 0 {
+		b.mu.Unlock()
+		ctx, cancel := context.WithTimeout(b.ctx, 5*time.Second)
+		defer cancel()
+		b.logout(ctx, true) // best effort
```
Comment (mihaip) [inline ipn/ipnlocal/local.go:355]: May as well use `LogoutSync` so the reader doesn't have to figure out what the boolean arg does?

File: cmd/tailscaled/tailscaled.go [hunk 4]
```diff
@@ -437,6 +436,7 @@ func run() error {
 	if err != nil {
 		return fmt.Errorf("safesocket.Listen: %v", err)
 	}
+	defer dialer.Close()
```
Comment (mihaip) [inline cmd/tailscaled/tailscaled.go:439]: Out of curiosity, why this change?

**Review by mihaip — APPROVED** [review]

File: ipn/ipnlocal/local.go (hunk 3 above)
Comment (maisem) [reply ipn/ipnlocal/local.go]: Good idea, done.

File: cmd/tailscaled/tailscaled.go (hunk 4 above)
Comment (maisem) [reply cmd/tailscaled/tailscaled.go:439]: It was being closed too early, which resulted in the call to LogoutSync not being able to make the network request to control.
